@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,524 @@
+        camera: THREE.Camera,
+        projection: Projection
+    ): { near: number; far: number } {
+        // Tilt angle and frustum volume will be taken into consideration only for non-spherical
+        // projections, otherwise take super class values.
+        if (projection.type !== ProjectionType.Spherical) {
+            const clipPlanes = { near: 0.0, far: 0.0 };
+            const cameraTilt = this.getCameraTiltAngle(camera, projection);
+            const lookAtDist = this.getCameraLookAtDistance(camera, projection);
+            // Generally near/far planes are set to keep look at distance, then
+            // margins are applied. Here margins are meant to be defined as distance
+            // along the ground normal vector thus during camera tilt they need to
+            // be projected on the eye vector:
+            // actualMargin = margin / groundNormal.dot(eyeVec)
+            // Assuming that tilt angle defined relative to actual ground normal, we have:
+            const cameraEyeDotGround = Math.cos(cameraTilt);
+            clipPlanes.near = lookAtDist - this.nearMargin / cameraEyeDotGround;
+            clipPlanes.far = lookAtDist + this.farMargin / cameraEyeDotGround;
+
+            // Correct cliping planse distace for the top/bottom frustum planes (edges).
+            // If we deal with perspective camera type, this step would not be required
+            // for orthografic projections.
+            if (camera.type === "PerspectiveCamera") {
+                const frustumGroundProj = this.getFrustumGroundIntersectionDist(
+                    camera as THREE.PerspectiveCamera,
+                    projection
+                );
+                // Looking on the graph presented in getFrustumGroundIntersectionDist() method
+                // we may, calculate far plane extension required in order to preserve look at
+                // based distance at any tilt angle for frustum edges, the formulas are
+                // presented below:
+                // cos(90 - tilt) = farPlaneExt / frustumIntersectTop
+                // farPlaneExt = cos(90 - tilt) * frustumIntersectTop
+                const tiltAngleRestCos = Math.cos(Math.PI / 2 - cameraTilt);
+                clipPlanes.far += frustumGroundProj.top * tiltAngleRestCos;
+                // Similar formulas may be derived for near plane distance:
+                // cos(90 - tilt) = nearPlaneExt / frustumIntersectBottom
+                // nearPlaneExt = frustumIntersectBottom * cos(90 - tilt)
+                clipPlanes.near -= frustumGroundProj.bottom * tiltAngleRestCos;
+            }
+            // Clamp values to constraints.
+            clipPlanes.near = Math.max(clipPlanes.near, this.nearMin);
+            clipPlanes.far = Math.min(clipPlanes.far, clipPlanes.near + this.planesMaxDistance);
+            return clipPlanes;
+        } else {
+            // Fallback to base implemenation case we don't yet have support for camera
```

Comment:
?

Reply:
It will be the same Jira ticket, although next commit. The sub-tasks were not created, although the entire ticket seems to be quite big, because it includes much of refactoring work, improving fixed planes evaluators, adding tilt based (planar & spherical projections) evaluators.